src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +137,132 @@ private void logTargetPath() {
         PathPlannerLogging.setLogTargetPoseCallback((pose) -> Logger.recordOutput("TargetPPPose", pose));
     }
 
+    /**
+     * Sets the estimated pose from the odometry at the given timestamp.
+     *
+     * @param swerveModulePositions the positions of each swerve module
+     * @param gyroHeading           the heading of the gyro
+     * @param timestamp             the timestamp of the odometry observation
+     */
+    private void addOdometryObservation(SwerveModulePosition[] swerveModulePositions, Rotation2d gyroHeading, double timestamp) {
+        final Twist2d newOdometryPoseDifference = calculateNewOdometryPoseDifference(swerveModulePositions, gyroHeading);
+        updateRobotPosesFromNewOdometryPoseDifference(newOdometryPoseDifference, timestamp);
+
+        updateOdometryPositions(swerveModulePositions, gyroHeading);
+    }
+
+    /**
+     * Updates the estimated pose from the cameras.
+     */
     private void updateFromVision() {
-        getViableVisionObservations().stream()
-                .sorted(Comparator.comparingDouble(PoseEstimator6328.VisionObservation::timestamp))
-                .forEach(poseEstimator6328::addVisionObservation);
+        Arrays.stream(aprilTagCameras).forEach(AprilTagCamera::update);
+
+        final AprilTagCamera[] newResultCameras = Arrays.stream(aprilTagCameras)
+                .filter(AprilTagCamera::hasNewResult)
+                .toArray(AprilTagCamera[]::new);
```

Comment:
I'd use normal `for`s, and separate each part into a function

Reply:
Why? This makes everything a lot simpler

Reply:
More efficient and readable

Reply:
The most efficient way to filter though is with a stream

Reply:
Pretty much by definition, a stream is less efficient.
https://www.linkedin.com/pulse/java-streams-vs-loops-why-arent-always-better-choice-pavan-pothuganti-wjcjc#:~:text=When%20it%20comes%20to%20performance,they%20might%20slow%20things%20down.

I'd use streams in cases where it's mandatory or just for a quick-n-dirty code